op-e2e/actions/op_program_test.go

```diff
@@ -0,0 +1,270 @@
+type L2FaultProofEnv struct {
+	log       log.Logger
+	batcher   *L2Batcher
+	sequencer *L2Sequencer
+	engine    *L2Engine
+	engCl     *sources.EngineClient
+	sd        *e2eutils.SetupData
+	dp        *e2eutils.DeployParams
+	miner     *L1Miner
+	alice     *CrossLayerUser
+}
+
+func NewL2FaultProofEnv(t Testing, tp *e2eutils.TestParams, dp *e2eutils.DeployParams, batcherCfg *BatcherCfg) *L2FaultProofEnv {
+	log := testlog.Logger(t, log.LvlDebug)
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+
+	miner, engine, sequencer := setupSequencerTest(t, sd, log)
+	miner.ActL1SetFeeRecipient(common.Address{0xCA, 0xFE, 0xBA, 0xBE})
+	sequencer.ActL2PipelineFull(t)
+	engCl := engine.EngineClient(t, sd.RollupCfg)
+
+	// Set the batcher key to the secret key of the batcher
+	batcherCfg.BatcherKey = dp.Secrets.Batcher
+	batcher := NewL2Batcher(log, sd.RollupCfg, batcherCfg, sequencer.RollupClient(), miner.EthClient(), engine.EthClient(), engCl)
+
+	addresses := e2eutils.CollectAddresses(sd, dp)
+	cl := engine.EthClient()
+	l2UserEnv := &BasicUserEnv[*L2Bindings]{
+		EthCl:          cl,
+		Signer:         types.LatestSigner(sd.L2Cfg.Config),
+		AddressCorpora: addresses,
+		Bindings:       NewL2Bindings(t, cl, engine.GethClient()),
+	}
+	alice := NewCrossLayerUser(log, dp.Secrets.Alice, rand.New(rand.NewSource(0xa57b)))
+	alice.L2.SetUserEnv(l2UserEnv)
+
+	return &L2FaultProofEnv{
+		log:       log,
+		batcher:   batcher,
+		sequencer: sequencer,
+		engine:    engine,
+		engCl:     engCl,
+		sd:        sd,
+		dp:        dp,
+		miner:     miner,
+		alice:     alice,
+	}
+}
+
+type TestParam func(p *e2eutils.TestParams)
+
+func NewTestParams(params ...TestParam) *e2eutils.TestParams {
+	dfault := defaultRollupTestParams
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type DeployParam func(p *e2eutils.DeployParams)
+
+func NewDeployParams(t Testing, params ...DeployParam) *e2eutils.DeployParams {
+	dfault := e2eutils.MakeDeployParams(t, NewTestParams())
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type BatcherCfgParam func(c *BatcherCfg)
+
+func NewBatcherCfg(params ...BatcherCfgParam) *BatcherCfg {
+	dfault := &BatcherCfg{
+		MinL1TxSize:          0,
+		MaxL1TxSize:          128_000,
+		DataAvailabilityType: batcherFlags.BlobsType,
+	}
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type OpProgramCfgParam func(p *config.Config)
+
+func NewOpProgramCfg(
+	t Testing,
+	env *L2FaultProofEnv,
+	l1Head common.Hash,
+	l2Head common.Hash,
+	l2OutputRoot common.Hash,
+	l2Claim common.Hash,
+	l2ClaimBlockNum uint64,
+	params ...OpProgramCfgParam,
+) *config.Config {
+	dfault := config.NewConfig(env.sd.RollupCfg, env.sd.L2Cfg.Config, l1Head, l2Head, l2OutputRoot, l2Claim, l2ClaimBlockNum)
+
+	// Set up in-process L1 sources
+	dfault.L1ProcessSource = env.miner.L1Client(t, env.sd.RollupCfg)
+	dfault.L1BeaconProcessSource = env.miner.blobStore
```

Comment:
hmm, I've been desperately trying to avoid putting code in the config objects - it makes them so hard to work with and mixes config and implementation. This might be the best option but I wonder if there's a cleaner way to inject them as part of actually creating the program...  Or maybe I just need to live with this being the pattern typically used in golang even though I don't like it.

Not a blocking objection, just noting it really. I can have a play and see what options there are.

Reply:
Yeah, it's not very pretty. One other option that we could do is pull the prefetcher's sources up higher. The reason it's in the config is because `makePrefetcher` is nested pretty deep into the entrypoint for the native host/client run, but if we moved that up we could just pass these sources directly into the invocation?

Reply:
I do like the idea of just passing in a prefetcher rather than sources. It's one input instead of three and way more flexible.  Just not sure what it will take to achieve until I play with the code. I have an inkling that it would be a nice pattern to have `FaultProofProgram` use the options patterns to let you pass in specific components that are then used - possibly as initializers if they have dependencies on other components.  So you'd get something like:

```
host.FaultProofProgram(cox, env.log, cfg, WithPrefetcher(prefetcher))
```